Lib/shutil.py

```diff
@@ -52,6 +52,7 @@
 COPY_BUFSIZE = 1024 * 1024 if _WINDOWS else 64 * 1024
 _USE_CP_SENDFILE = hasattr(os, "sendfile") and sys.platform.startswith("linux")
 _HAS_FCOPYFILE = posix and hasattr(posix, "_fcopyfile")  # macOS
+_WIN_DEFAULT_PATHEXT = ".COM;.EXE;.BAT;.CMD;.VBS;.JS;.WS;.MSC"
```

Comment:
Add a comment like "default in Windows 10 CMD", so we know where this comes from. 

Back in NT 4, CMD's default was just `".COM;.EXE;.BAT;.CMD"`. Windows 2000 (NT 5) added support for additional scripting languages (.VBS, .JS, .WS) and MMC saved console files (.MSC). In Windows 7+, the default *installed value* is `".COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC"`. This includes encoded scripts (.VBE, .JSE) and mixed-language scripts (.WSF, .WSH). CMD's current default includes ".WS", but this is not a supported Windows Scripting file extension nowadays. Which do you think is better -- CMD's default value (used by `os.system` and subprocess with `shell=True`), which is a bit outdated, or the default installed value?

Reply:
@eryksun I gave it some thought and I seem to have a hard time coming to a conclusion. What would the user of the language expect? From a language point of view, the expectation could be that we are aligning with the `os` and `subprocess` module here. On the other hand, people might expect we honor the default installed value of Windows. Yet, it would mean we have to look at them with every Windows release and re-check if we are still aligned (I have no sense of how often they change). I personally would go with the CMD defaults for now, even though I have a hard time finding a proper argument why one is more suited than the other.

For now, I just added the comment. Feel free to resolve this conversation if you agree and if not, I would be happy to hear more insights.

Thank you for your thorough explanation 😃

Reply:
I prefer the shorter default, mostly to avoid having the script/malware files in there, but then if someone has cleared the variable they probably had a good reason and it may be best to treat it as empty and not automatically handle any extensions in `which`.

`os.system` and subprocess with `shell=True` and explicitly platform-specific behaviour, so there's no good reason to try and imitate them. But some quick playing around gave me this:

```
C:\> set PATHEXT=
C:\> set PATHEXT
Environment variable PATHEXT not defined
C:\>where cmd
INFO: Could not find files for the given pattern(s).
C:\> cmd
Microsoft Windows [Version 10.0.18363.778]
(c) 2019 Microsoft Corporation. All rights reserved.
```

Oh weird, it resolved `cmd.exe` anyway. Still in the child process:

```
C:\>set PATHEXT
PATHEXT=.COM;.EXE;.BAT;.CMD;.VBS;.JS;.WS;.MSC
```

Okay, it reset it to something different from how it started. Let's exit and try something else:

```
C:\> set PATHEXT=;
C:\> cmd
'cmd' is not recognized as an internal or external command,
operable program or batch file.
C:\> cmd.exe
Microsoft Windows [Version 10.0.18363.778]
(c) 2019 Microsoft Corporation. All rights reserved.

C:\>set PATHEXT
PATHEXT=;
```

Okay, properly inherited when it's not empty. Let's try Powershell:

```
C:\> set PATHEXT=
C:\> powershell
Windows PowerShell
Copyright (C) Microsoft Corporation. All rights reserved.
PS C:\> $env:PATHEXT
.CPL
```

Weird, it set `.CPL`.

So I think the best we can assume is that everything goes weird if it's empty. In that case, I vote for treating an empty variable as empty, and not adding any extensions.

Reply:
> So I think the best we can assume is that everything goes weird if it's empty. In that case, I vote for treating an empty variable as empty, and not adding any extensions.

The intent of my suggestion was to harmonize `shutil.which` with CMD, as is used by `os.system` and subprocess with `shell=True`. So running `subprocess.run(shutil.which(basename))` should find and run the same file as `subprocess.run(basename, shell=True)` -- assuming the file is found, and ignoring edge cases due to differences in how it searches. 

At the very least, `shutil.which` isn't really useful on Windows if it doesn't have a basic default of ".COM;.EXE;.BAT;.CMD", which was CMD's default back in NT 4 (1996). Those four extensions all execute directly via `CreateProcessW`.